packages/jsii/lib/transforms/deprecated-warnings-transformer.ts

```diff
@@ -0,0 +1,413 @@
+          ElementType.PARAMETER,
+          this.moduleName,
+        )
+      : undefined; // There is no deprecated type in its heritage chain
+  }
+
+  private getParameterWarnings(
+    toProcess: ts.Node[],
+    result: Warning[] = [],
+  ): Warning[] {
+    if (toProcess.length === 0) {
+      return result;
+    }
+
+    const node = toProcess[0];
+    const warning = this.getWarningForHeritage(node);
+    const warnings: Warning[] = warning != null ? [warning] : [];
+    if (isDeprecated(node)) {
+      warnings.push(getWarning(node, ElementType.PARAMETER, this.moduleName));
+    }
+    const type = this.typeChecker.getTypeAtLocation(node);
+    const declaration = type.symbol?.declarations[0];
+    let newBatch: ts.Node[] = [];
+    // In some cases (e.g., enums) the declaration is the node itself. Not including them twice here
+    if (declaration != null && declaration !== node) {
+      if (isDeprecated(declaration)) {
+        warnings.push(
+          getWarning(declaration, ElementType.PARAMETER, this.moduleName),
+        );
+      }
+      if (
+        ts.isInterfaceDeclaration(declaration) ||
+        ts.isClassDeclaration(node)
+      ) {
+        newBatch = type.getProperties().map((p) => p.declarations[0]);
+      }
+      if (ts.isEnumDeclaration(declaration)) {
+        newBatch = [...declaration.members];
+      }
+    }
+    return this.getParameterWarnings(
+      toProcess.slice(1).concat(newBatch),
+      result.concat(warnings),
+    );
+  }
+
+  private createWarningStatements(): ts.Statement[] {
+    return [...this.warnings].flatMap((w) => this.createWarningStatement(w));
+  }
+
+  // TODO Rename this method
+  private findDeprecatedInTheTypeHierarchy(
+    toProcess: ts.Type[],
+  ): ts.Type | undefined {
+    if (toProcess.length === 0) {
+      return undefined;
+    }
+
+    const type = toProcess[0];
+    if (this.deprecatedTypes.includes(type)) {
+      return type;
+    }
+
+    const node = this.index.get(
+      this.typeChecker.getFullyQualifiedName(type.symbol),
+    );
+
+    // This type was not declared in the JSII assembly. Skipping.
+    if (node == null) {
+      return undefined;
+    }
+
+    if (!ts.isClassDeclaration(node) && !ts.isInterfaceDeclaration(node)) {
+      throw new Error('Node should be a class or interface declaration.');
+    }
+
+    const newBatch = toProcess.slice(1);
+    const heritageClauses = node.heritageClauses ?? [];
+    for (const clause of heritageClauses) {
+      for (const expression of clause.types) {
+        const type = this.typeChecker.getTypeAtLocation(expression);
+        if (!toProcess.includes(type)) {
+          newBatch.push(type);
+        }
+      }
+    }
+
+    return this.findDeprecatedInTheTypeHierarchy(newBatch);
+  }
+
+  private createWarningStatement(warning: Warning): ts.Statement[] {
+    const message = `${warning?.elementName} is deprecated.\n  ${warning?.message}\n  This API will be removed in the net major release.`;
```

Comment:
Why all the `?.`? `warning` is a required parameter here...

```suggestion
    const message = `${warning.elementName} is deprecated.\n  ${warning.message}\n  This API will be removed in the next major release.`;
```

Reply:
Yes, they shouldn't be there. Thanks for catching this.